third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,242 @@
+        let num_ty_params = self.mod_env.get_function(fun_id).get_type_parameter_count() as u16;
+        (0..num_ty_params).map(Type::TypeParameter).collect_vec()
+    }
+
+    /// Generates the initial callers chain for the given function,
+    /// which is the given function initialized with generic type parameters
+    fn gen_init_callers_chain(&self, fun_id: FunId) -> Vec<(Loc, QualifiedInstId<FunId>)> {
+        let insts = self.gen_generic_insts_for_fun(fun_id);
+        let root_caller = self.mod_env.get_id().qualified_inst(fun_id, insts);
+        let root_caller_loc = self.mod_env.get_function(fun_id).get_loc();
+        vec![(root_caller_loc, root_caller)]
+    }
+
+    /// Visits an expression and checks for cyclic type instantiations.
+    /// `insts`: the type parameters of the current expression
+    /// `callers_chain`: the chain of callers leading to the current expression
+    fn visit(
+        &self,
+        position: VisitorPosition,
+        e: &ExpData,
+        insts: &[Type],
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        use ExpData::*;
+        use VisitorPosition::*;
+        match (position, e) {
+            (Pre, Call(nid, op, _)) => self.visit_call(nid, op, insts, callers_chain),
+            _ => true,
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`
+    fn visit_call(
+        &self,
+        nid: &NodeId,
+        op: &Operation,
+        insts: &[Type],
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        if let Operation::MoveFunction(mod_id, fun_id) = op {
+            let callee_uninst = mod_id.qualified_inst(*fun_id, self.get_inst(*nid));
+            let callee = callee_uninst.instantiate(insts);
+            if *mod_id != self.mod_env.get_id() || self.def_not_recursive(callee.to_qualified_id())
+            {
+                // skips if callee from another module (since there is no cyclic module dependency),
+                // or if the callee is not recursive
+                true
+            } else {
+                for (_, ancester_caller) in callers_chain.iter() {
+                    if ancester_caller.to_qualified_id() == callee.to_qualified_id() {
+                        // we are checking for the root caller
+                        let (_, checking_for) = &callers_chain[0];
+                        if checking_for.to_qualified_id() != callee.to_qualified_id() {
+                            // check and report diagnostics when `callee` is checked
+                            // this happens when root caller `f` calls `g` which then calls `g` itself
+                            return true;
+                        } else {
+                            #[allow(clippy::collapsible_else_if)]
+                            if let Some(_ty_param) = callee
```

Comment:
We could use an explanatory comment here, I think.  How does even a single type parameter indicate an error?  Couldn't there have been an abstract type parameter in the original call?

Reply:
The method is called ty_*properly*_contains_ty_parameter and a single type parameter will not be included.